[BUG] [Searchable Snapshot] SpecialPermission error when using a scripted query

**Describe the bug**
A clear and concise description of what the bug is.
When performing a scripted query on a remote index, a security exception is thrown. 
This occurs on 2.7.0 snapshot and also using released 2.5.0. 

Seems related to how OS/ES handles evaluated permissions. 

**To Reproduce**
Steps to reproduce the behavior:
1. Restore an index where `storage_type` is set to remote_snapshot
2. Perform a scripted search using `painless`
3. A security exception is thrown. On a normal index, no exception occurs. 

**Expected behavior**
A clear and concise description of what you expected to happen.

**Plugins**
Please list all plugins currently enabled.

**Screenshots**
If applicable, add screenshots to help explain your problem.
```
 Caused by: java.lang.SecurityException: access denied ("org.opensearch.SpecialPermission" "*")
 	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:485) ~[?:?]
 	at java.security.AccessController.checkPermission(AccessController.java:1068) ~[?:?]
 	at java.lang.SecurityManager.checkPermission(SecurityManager.java:416) ~[?:?]
 	at org.opensearch.SpecialPermission.check(SpecialPermission.java:104) ~[opensearch-2.5.0.jar:2.5.0]
 	at org.opensearch.repositories.azure.SocketAccess.doPrivilegedException(SocketAccess.java:68) ~[?:?]
 	at org.opensearch.repositories.azure.AzureBlobStore.getInputStream(AzureBlobStore.java:282) ~[?:?]
 	at org.opensearch.repositories.azure.AzureBlobContainer.openInputStream(AzureBlobContainer.java:104) ~[?:?]
 	at org.opensearch.repositories.azure.AzureBlobContainer.readBlob(AzureBlobContainer.java:122) ~[?:?]
 	at org.opensearch.index.store.remote.utils.TransferManager.fetchBlob(TransferManager.java:59) ~[opensearch-2.5.0.jar:2.5.0]
 	at org.opensearch.index.store.remote.file.OnDemandBlockSnapshotIndexInput.fetchBlock(OnDemandBlockSnapshotIndexInput.java:151) ~[opensearch-2.5.0.jar:2.5.0]
 	at org.opensearch.index.store.remote.file.OnDemandBlockIndexInput.demandBlock(OnDemandBlockIndexInput.java:347) ~[opensearch-2.5.0.jar:2.5.0]
 	at org.opensearch.index.store.remote.file.OnDemandBlockIndexInput.seekInternal(OnDemandBlockIndexInput.java:318) ~[opensearch-2.5.0.jar:2.5.0]
 	at org.opensearch.index.store.remote.file.OnDemandBlockIndexInput.seek(OnDemandBlockIndexInput.java:216) ~[opensearch-2.5.0.jar:2.5.0]
 	at org.opensearch.index.store.remote.file.OnDemandBlockSnapshotIndexInput.seek(OnDemandBlockSnapshotIndexInput.java:30) ~[opensearch-2.5.0.jar:2.5.0]
                    
```
**Host/Environment (please complete the following information):**
 - OS: [e.g. iOS] Ubuntu
 - Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

--- a/server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java
+++ b/server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java
@@ -13,6 +13,7 @@ import org.apache.logging.log4j.Logger;
 import org.apache.lucene.store.IOContext;
 import org.apache.lucene.store.IndexInput;
 import org.opensearch.common.blobstore.BlobContainer;
+import org.opensearch.index.store.remote.filecache.CachedIndexInput;
 import org.opensearch.index.store.remote.filecache.FileCache;
 import org.opensearch.index.store.remote.filecache.FileCachedIndexInput;
 
@@ -20,8 +21,11 @@ import java.io.BufferedOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.io.UncheckedIOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 
 /**
  * This acts as entry point to fetch {@link BlobFetchRequest} and return actual {@link IndexInput}. Utilizes the BlobContainer interface to
@@ -58,15 +62,17 @@ public class TransferManager {
                 }
             } else {
                 if (cachedIndexInput.isClosed()) {
-                    // if it's already in the file cache, but closed, open it and replace the original one
-                    try {
-                        final IndexInput luceneIndexInput = blobFetchRequest.getDirectory()
-                            .openInput(blobFetchRequest.getFileName(), IOContext.READ);
-                        return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), luceneIndexInput);
-                    } catch (IOException e) {
-                        logger.warn("Failed to open existing file for " + blobFetchRequest.getFilePath(), e);
-                        return null;
-                    }
+                    return AccessController.doPrivileged((PrivilegedAction<CachedIndexInput>) () -> {
+                        // if it's already in the file cache, but closed, open it and replace the original one
+                        try {
+                            final IndexInput luceneIndexInput = blobFetchRequest.getDirectory()
+                                .openInput(blobFetchRequest.getFileName(), IOContext.READ);
+                            return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), luceneIndexInput);
+                        } catch (IOException e) {
+                            logger.warn("Failed to open existing file for " + blobFetchRequest.getFilePath(), e);
+                            return null;
+                        }
+                    });
                 }
                 // already in the cache and ready to be used (open)
                 return cachedIndexInput;
@@ -88,17 +94,27 @@ public class TransferManager {
     }
 
     private IndexInput downloadBlockLocally(BlobFetchRequest blobFetchRequest) throws IOException {
-        try (
-            InputStream snapshotFileInputStream = blobContainer.readBlob(
-                blobFetchRequest.getBlobName(),
-                blobFetchRequest.getPosition(),
-                blobFetchRequest.getLength()
-            );
-            OutputStream fileOutputStream = Files.newOutputStream(blobFetchRequest.getFilePath());
-            OutputStream localFileOutputStream = new BufferedOutputStream(fileOutputStream);
-        ) {
-            snapshotFileInputStream.transferTo(localFileOutputStream);
-        }
-        return blobFetchRequest.getDirectory().openInput(blobFetchRequest.getFileName(), IOContext.READ);
+        // We need to do a privileged action here in order to fetch from remote
+        // and write to the local file cache in case this is invoked as a side
+        // effect of a plugin (such as a scripted search) that doesn't have the
+        // necessary permissions.
+        return AccessController.doPrivileged((PrivilegedAction<? extends IndexInput>) () -> {
+            try {
+                try (
+                    InputStream snapshotFileInputStream = blobContainer.readBlob(
+                        blobFetchRequest.getBlobName(),
+                        blobFetchRequest.getPosition(),
+                        blobFetchRequest.getLength()
+                    );
+                    OutputStream fileOutputStream = Files.newOutputStream(blobFetchRequest.getFilePath());
+                    OutputStream localFileOutputStream = new BufferedOutputStream(fileOutputStream);
+                ) {
+                    snapshotFileInputStream.transferTo(localFileOutputStream);
+                }
+                return blobFetchRequest.getDirectory().openInput(blobFetchRequest.getFileName(), IOContext.READ);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
     }
 }

--- a/server/src/test/java/org/opensearch/index/store/remote/utils/TransferManagerTests.java
+++ b/server/src/test/java/org/opensearch/index/store/remote/utils/TransferManagerTests.java
@@ -10,6 +10,7 @@ package org.opensearch.index.store.remote.utils;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.ExecutorService;
@@ -157,7 +158,7 @@ public class TransferManagerTests extends OpenSearchTestCase {
     public void testDownloadFails() throws Exception {
         doThrow(new IOException("Expected test exception")).when(blobContainer).readBlob(eq("failure-blob"), anyLong(), anyLong());
         expectThrows(
-            IOException.class,
+            UncheckedIOException.class,
             () -> transferManager.fetchBlob(
                 BlobFetchRequest.builder()
                     .blobName("failure-blob")

